### PR TITLE
[GITHUB]  Badge for total stars of an user / org

### PR DESCRIPTION
**Pull request by hemantsonu20**
* Two different graphql query needed for this badge one for `user` account and another for `organization` account.
* Limit set to 200 repos. i.e. only upto 200 most starred repos will be considered.
* Added a styling to highlight texts in documentation. Any text can be highlighted similar to `this text` in documentation.
* Added a `transformGraphqlErrors` method to dynamically create error with pretty message. For example while querying for a `user`, a custom error `user not found` will be thrown.

### CODE REVIEW

File: services/github/github-total-star.service.js
Comment (calebcartwright) [inline services/github/github-total-star.service.js:40]: No need to add the `.required()` here, as it's already included by `nonNegativeInteger` (`optionalNonNegativeInteger` is the optional version of this)
```suggestion
        totalCount: nonNegativeInteger,
```

File: services/github/github-total-star.service.js
Comment (calebcartwright) [inline services/github/github-total-star.service.js:152]: I don't think `users` is the right noun to have here (see our [badge url conventions](https://github.com/badges/shields/blob/master/doc/badge-urls.md#badge-url-conventions) for more info). Something like `stars` or `total-stars` would be better.

Personally I think `stars` would suffice and be a better route match with the existing repo stars, and there's no concerns about route conflicts (the other requires an owner and repo), and there's query params used to toggle between an org and user to differentiate between those two

File: services/github/github-total-star.service.js
Comment (calebcartwright) [inline services/github/github-total-star.service.js:143]: I'm inclined to remove the `Total` word from here, especially given the potential caveats. It's really not the total star count for a User nor an Org, just a different (and larger but potentially not complete) scope than the `Stars` badge for a single repo

File: services/github/github-total-star.service.js
Comment (calebcartwright) [inline services/github/github-total-star.service.js:171]: This example doesn't work for me on the review app, nor do any other users i've tried

https://shields-staging-pr-5507.herokuapp.com/github/users/chris48s/total-stars?style=social


https://shields-staging-pr-5507.herokuapp.com/github/users/chris48s/total-stars?affiliations=OWNER%2CCOLLABORATOR&style=social


https://shields-staging-pr-5507.herokuapp.com/github/users/sindresorhus/total-stars?affiliations=OWNER%2CCOLLABORATOR&style=social

File: services/github/github-total-star.service.js
Comment (calebcartwright) [inline services/github/github-total-star.service.js:185]: This is working for me but it was running on the slow side in the review app. May just be a fluke, but should probably poke around a bit to make sure this badge is providing quickly enough to meet the github/camo window

https://shields-staging-pr-5507.herokuapp.com/github/users/badges/total-stars?org&style=social

File: services/github/github-total-star.tester.js
Comment (calebcartwright) [inline services/github/github-total-star.tester.js:10]: Would be good to just use the `nonNegativeInteger` for consistency in these

File: services/github/github-total-star.tester.js
Comment (calebcartwright) [inline services/github/github-total-star.tester.js:50]: Let's use `isMetric` validator here instead

File: services/github/github-total-star.tester.js
Comment (calebcartwright) [inline services/github/github-total-star.tester.js:58]: Let's use `isMetric` validator here instead

File: services/github/github-total-star.service.js
Comment (calebcartwright) [inline services/github/github-total-star.service.js:141]:
```suggestion
module.exports = class GithubTotalStarService extends GithubAuthV4Service {
```

**Review by calebcartwright — COMMENTED** [review]
Thank you for the PR @hemantsonu20! I've taken an initial pass through this and left some feedback and requested changes for your review.

The biggest challenge I'm seeing at the moment is that the badges for Users doesn't seem to be working

File: frontend/components/markup-modal/markup-modal-content.tsx
Comment (hemantsonu20) [reply frontend/components/markup-modal/markup-modal-content.tsx]: I removed this change, and replaced my documentation with `<code></code>` instead of `<mark></mark>`. It doesn't appears as I wanted like `this` but ok for now.

File: services/github/github-helpers.js
Comment (hemantsonu20) [reply services/github/github-helpers.js]: Done.

File: services/github/github-total-star.service.js
Comment (hemantsonu20) [reply services/github/github-total-star.service.js]: Its now `Stars` only.

File: services/github/github-total-star.service.js
Comment (hemantsonu20) [reply services/github/github-total-star.service.js]: It should work now. A little confusing fields `affiliations` and `ownerAffiliations` were there in graphql. Github doc didn't explained which one to use nd when. When I used `ownerAffiliations`, it worked for my account as well as other account.
`affiliations` allowed filtering on my account only, and for other accounts, without throwing any errors, it retuerned zero repos.

File: services/github/github-total-star.service.js
Comment (hemantsonu20) [reply services/github/github-total-star.service.js]: did no changes for this, but verified only two sequential call is getting fired.

File: services/github/github-total-star.tester.js
Comment (hemantsonu20) [reply services/github/github-total-star.tester.js]: I used `isMetric` validator here, which makes more sense.

File: services/github/github-total-star.tester.js
Comment (hemantsonu20) [reply services/github/github-total-star.tester.js]: Done

File: services/github/github-total-star.service.js
Comment (hemantsonu20) [reply services/github/github-total-star.service.js]: Done

File: services/github/github-total-star.tester.js
Comment (hemantsonu20) [reply services/github/github-total-star.tester.js]: Done

File: services/github/github-total-star.service.js
Comment (hemantsonu20) [reply services/github/github-total-star.service.js]: Done

File: services/github/github-total-star.service.js
Comment (hemantsonu20) [reply services/github/github-total-star.service.js]: Done

File: services/github/github-total-star.service.js
Comment (calebcartwright) [inline services/github/github-total-star.service.js:155]: Let's use the new static props that we're converting to so that we can avoid adding rework

```suggestion
  static defaultLabel = 'Stars'
  static category = 'social'
  static route = {
      base: 'github/stars',
      pattern: ':user',
      queryParamSchema,
  }
```

File: services/github/github-total-star.service.js
Comment (calebcartwright) [inline services/github/github-total-star.service.js:186]:
```suggestion
  static examples = [
    {
      title: "Github User's stars",
      namedParams: {
        user: 'chris48s',
      },
      queryParams: { affiliations: 'OWNER,COLLABORATOR' },
      staticPreview: {
        label: this.defaultLabel,
        message: 54,
        style: 'social',
      },
      documentation: userDocumentation,
    },
    {
      title: "Github Org's stars",
      namedParams: {
        user: 'badges',
      },
      queryParams: { org: null },
      staticPreview: {
        label: this.defaultLabel,
        message: metric(7000),
        style: 'social',
      },
      documentation: orgDocumentation,
    },
  ]
```

File: services/github/github-total-star.service.js
Comment (calebcartwright) [inline services/github/github-total-star.service.js:193]:
```suggestion
  static defaultBadgeData = {  label: this.defaultLabel, namedLogo: 'github'  }
```

File: services/github/github-total-star.service.js
Comment (calebcartwright) [inline services/github/github-total-star.service.js:15]:
```suggestion
const customDocumentation = `This badge takes into account up to <code>${MAX_REPO_LIMIT}</code> of the most starred repositories of given user / org.`
```

File: services/github/github-total-star.service.js
Comment (calebcartwright) [inline services/github/github-total-star.service.js:23]:
```suggestion
  Default value is <code>OWNER</code>. See the explanation of these values <a href="https://docs.github.com/en/graphql/reference/enums#repositoryaffiliation">here</a>.
```

File: services/github/github-total-star.service.js
Comment (calebcartwright) [inline services/github/github-total-star.service.js:21]:
```suggestion
  2. <code>affiliations</code> query param accepts three values (must be UPPER case) <code>OWNER</code>, <code>COLLABORATOR</code>, <code>ORGANIZATION_MEMBER</code>.
```

File: services/github/github-total-star.service.js
Comment (calebcartwright) [inline services/github/github-total-star.service.js:22]:
```suggestion
  One can pass comma separated combinations of these values (no spaces) e.g. <code>OWNER,COLLABORATOR</code> or <code>OWNER,COLLABORATOR,ORGANIZATION_MEMBER</code>.
```

**Review by calebcartwright — COMMENTED** [review]
Thanks for the updates! Few minor closing items from me, but otherwise looks good to go

**Comment by chris48s** [comment]
Hello. @calebcartwright  has requested a second review on this one. No concerns on efficiency: as it stands, each query is only one rate limit point. I haven't really had a chance to do a full read-through of the code (I'm currently away from home and its a bit inconvenient) but based on a quick play I think there is one big issue that could be improved here:

This `?org` URL is a bit annoying. Its nice that `/stars/:user/:repo` is a badge and `/stars/:user` is too. Its annoying that `/stars/:org/:repo` works but the corresponding badge has to be `/stars/:org?org` :(
In principle I think we don't need to do this. We can issue

```gql
query {
  user (login: "badges"){
    //...
  }
  organization(login: "badges") {
    //...
  }
}
```
in one call and then work out from the response whether "badges" was a user or an org (or neither). I've tested this in the GH graphql explorer and it

a) works
and
b) still only seems to cost 1 rate limit point

but I'm not 100% sure how `GithubAuthV4Service` or `BaseGraphqlService` handles a response that contains both an error and valid data. It might need a tweak there (or it might just work depending on how clever Chris-from-the-past was :wink:  ). If at all possible, I think we should do this instead of making the user pass `?org`. It would be much nicer if:

- `/stars/badges/shields`
- `/stars/badges`
- `/stars/chris48s/geojson-rewind`
- `/stars/chris48s`

all _just work_ ™ and it is better to do a bit more work on our side to make that happen.

---

I'll leave a few more minor comments inline. I should be able to circle back and look at this in more detail in a few days.

Thanks for your efforts on this so far both @hemantsonu20  and @calebcartwright might take a while to get it right, but we will get there :)

File: services/github/github-total-star.service.js
Comment (chris48s) [inline services/github/github-total-star.service.js:152]: We style as GitHub (with a capital H) not Github - may be other occurrences

Comment (chris48s) [reply services/github/github-total-star.service.js:152]: Also, can we also modify the description of the existing GH stars badge so it is now clear that one is stars for a repo (at the moment that one is just labelled "GitHub stars" which is now confusing).

File: services/github/github-total-star.service.js
Comment (chris48s) [inline services/github/github-total-star.service.js:141]: "stars" should be all lowercase: https://github.com/badges/shields/blob/master/CONTRIBUTING.md#badge-guidelines

File: services/github/github-total-star.service.js
Comment (chris48s) [inline services/github/github-total-star.service.js:239]: Can we optimise this a bit. If we have already found the first repo with zero stars while processing the first page we don't need to make the second request at all - we already know all the repos on page 2 will have 0 stars.

File: services/github/github-total-star.service.js
Comment (chris48s) [inline services/github/github-total-star.service.js:164]: We can do

```js
pattern: ':org',
namedParams: {
  org: 'badges',
},
```

here on this example so that the param is styled as `:org` in the front-end instead of `:user`, just to make it slightly clearer for users :)

File: core/base-service/base-graphql.js
Comment (chris48s) [inline core/base-service/base-graphql.js:49]: Can we specify the default value here so the param is marked as optional when we compile the docs, like this

```suggestion
   * @param {Function} [attrs.transformJson=data => data] Function which takes the raw json and transforms it before
```

(might need to re-run prettier after this for line lengths?)

**Review by chris48s — COMMENTED** [review]
Nice. Two minor comments and this is good to ship IMO

File: services/github/github-total-star.service.js
Comment (hemantsonu20) [reply services/github/github-total-star.service.js:164]: Earlier I thought same, but didn't find any hack for it. It's better now.

File: core/base-service/base-graphql.js
Comment (hemantsonu20) [reply core/base-service/base-graphql.js:49]: Still new to javascript world :)

**Review by chris48s — APPROVED** [review]
great first contribution

**Comment by calebcartwright** [comment]
Great job @hemantsonu20!

**Comment by hemantsonu20** [comment]
Thanks @calebcartwright @chris48s for detailed reviews and accepting the PR

**Pull request merged**
